worker/src/instagram.rs

```diff
@@ -0,0 +1,81 @@
+use crate::utils::extract_signature;
+use anyhow::{anyhow, Result};
+use reqwest::header::{HeaderMap, AUTHORIZATION};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct User {
+    pub username: String,
+    pub id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Post {
+    pub sig: String,
+    pub sig_target: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CaptionWrapper {
+    pub caption: String,
+    pub permalink: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PostWrapper {
+    pub data: Vec<PostId>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PostId {
+    pub id: String,
+}
+
+pub async fn retrieve_post(handle: String, token: String) -> Result<(String, String, String)> {
+    let url = format!(
+        "https://graph.instagram.com/me?fields=id,username&access_token={}",
+        token
+    );
+
+    let client = reqwest::Client::new();
+    let res: User = client.get(Url::parse(&url)?).send().await?.json().await?;
+
+    if handle != res.username {
+        return Err(anyhow!(format!(
+            "Token is not authorized for handle: {}",
+            handle
+        )));
+    }
+
+    let url = format!(
+        "https://graph.instagram.com/v11.0/{}/media?access_token={}",
+        res.id, token
+    );
+
+    let res: PostWrapper = client.get(Url::parse(&url)?).send().await?.json().await?;
+
+    let mut n = res.data.len();
+    if n == 0 {
+        return Err(anyhow!("No posts found for given user"));
+    } else if n > 100 {
+        n = 100;
+    }
+
+    let search = res.data.iter().take(n);
+
+    for data in search {
+        let url = format!(
+            "https://graph.instagram.com/{}?fields=caption,permalink&access_token={}",
+            data.id, token
+        );
+
+        let res: CaptionWrapper = client.get(Url::parse(&url)?).send().await?.json().await?;
+        match extract_signature(res.caption) {
+            Ok((sig_target, sig)) => return Ok((sig_target, sig, res.permalink.clone())),
+            Err(_) => {}
+        }
+    }
```

Comment:
Why don't we ask the user for the post URL? (I guess that's related to the worker doing the app authorisation.)

Reply:
Because there's no way to query the Instagram API based on the information in the post URL. It can't be de-referenced by TZP or the worker and contains no useful information in the path.

Reply:
Damn the slug id isn't a media id? :pensive: